FHIR: Data Dictionary and API

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Spec](https://docs.google.com/document/d/1q2ZSIuXxBenIUo3ROCpJ9lExPmE5huaDW5kBo4BaqVE/edit#heading=h.34p9d9xpfe1k)

This PR allows users to set up FHIR Resource mapping from Data Dictionary UI. 
1. Set up FHIR Resource Type by selecting from a list of supported Resource Types (Patient, DiagnosticReport, Observation)
2. Set up path (text) to a FHIR resource property for each Case Property

Once the mapping (case properties -> fhir resource properties) is done, the FHIR API will use it to create the JSON response.

This makes use of the pre-existing models added in https://github.com/dimagi/commcare-hq/pull/29100.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`DATA_DICTIONARY` & `FHIR_INTEGRATION`(New)

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
![Screenshot from 2021-04-02 15-20-21](https://user-images.githubusercontent.com/3864163/113405236-21379d80-93c7-11eb-9821-8a526877bf3d.png)


## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Tests were added for model and util functions. Tests for views are to be taken as a followup, though the full feature has been QA'ed

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-2634

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
All of the relevant code changes are behind the new feature flag.
Tested locally to make sure things are working well with/without feature flag turned on with new set of changes. But will be doing another pass for edge cases before merging.
QA was done on staging as well.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

There is a migration in this PR. There is definitely data on staging that was generated by QA, and it would be good if that could still be present. I believe if you just revert the PR, it would leave the table as it is.
There is no domain using it on any other environment and is a standalone feature as well. 
If any issue arises from this PR and it can't be fixed, it should be fine to revert this.

mkangia committed:
allow setting up fhir resource type and path

mkangia commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -97,6 +103,8 @@ hqDefine("data_dictionary/js/data_dictionary", [
                     data: {
                         'properties': JSON.stringify(postProperties),
+                        'fhir_resource_type': self.fhirResourceType(),
+                        'case_type': self.activeCaseType(),
```
Here we are creating the JSON to send for saving various case properties. Each case prop object also has the same "case_type" name added and then when processing the request, we iterate on each case prop and process it.
Now I need the case_type on backend before iterating the props passed to create fhir resource type. I could have read from one of the case properties passed but its not necessary that there will be any, unlikely but possible. So I have passed it separately in the request.

mkangia commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -47,6 +63,9 @@ <h3 data-bind="text: $root.activeCaseType()"></h3>
           <td></td>
           <td></td>
+          {% if fhir_integration_enabled %}
+            <td></td>
```
This just adds an empty td, in case needed to fill up for empty space in header.

kaapstorm commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -97,17 +115,35 @@ def data_dictionary_json(request, domain, case_type_name=None):
     errors = []
+    if fhir_resource_type and case_type:
+        case_type_obj = CaseType.objects.get(domain=domain, name=case_type)
+        try:
+            fhir_resource_type_obj = FHIRResourceType.objects.get(case_type=case_type_obj, domain=domain)
+        except FHIRResourceType.DoesNotExist:
+            fhir_resource_type_obj = FHIRResourceType(case_type=case_type_obj, domain=domain)
```
Django has a shortcut for this.
```suggestion
        fhir_resource_type_obj = FHIRResourceType.objects.get_or_create(
            domain=domain,
            case_type=case_type_obj,
        )
```

mkangia replied:
Ya, i was trying to avoid two SQL saves. Since I need to update the object's name and then save it anyway. If i was just creating the object, I would have used get_or_create.
Does that sound reasonable?

kaapstorm replied:
That does make sense.

(I made a mistake in my suggested change: `get_or_create()` returns a tuple, `obj, created`. You could use `created` to figure out if you need to update the existing instance, but I think what you've got here is neater.)

kaapstorm commented on corehq/motech/fhir/const.py:
```diff
@@ -2,3 +2,8 @@
 FHIR_VERSIONS = [
     (FHIR_VERSION_4_0_1, 'R4'),
 ]
+SUPPORTED_FHIR_RESOURCE_TYPES = [
```
Will this list always be small, for usability, or will it grow to include all FHIR resource types? (i.e. the return value of [FHIRResourceType.get_names()](https://github.com/dimagi/commcare-hq/pull/29100/files#diff-433e7032abf02a0c78231568078bff8181998607bf011edb22ea4b1d3018041fR59))

mkangia replied:
For the use case that I know of currently, I have limited it to these 3. I don't know if there is any harm in letting this list be all Resource types, so thought of limiting it for now.

kaapstorm commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -72,6 +91,15 @@ <h3 data-bind="text: $root.activeCaseType()"></h3>
           </td>
+          {% if fhir_integration_enabled %}
+          <td>
+            <textarea class="form-control vertical-resize" data-bind="
```
Could I argue in favor of a text input instead of a textarea? I'm hoping that the JSON path won't be super long.

mkangia replied:
absolutely, i thought these would be long. Updated.

mkangia committed:
use only text input

mkangia committed:
add fhir resource get view

mkangia committed:
add search view

mkangia committed:
enforce GET

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -36,7 +38,7 @@ hqDefine("data_dictionary/js/data_dictionary", [
     };
 
-    var propertyListItem = function (name, isGroup, groupName, caseType, dataType, description, deprecated) {
+    var propertyListItem = function (name, isGroup, groupName, caseType, dataType, description, fhirRsourcePropPath, deprecated) {
```
Optional: I'm a fan of passing `options` object when param lists get long and then using [assert_properties](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/assert_properties.js) to verify expected params are included. It also lets you distinguish between required and optional, which would be nice here where the fhir properties are their own flag.

mkangia replied:
That makes sense. Going to skip this for now @orangejenny . 
I tried that i would just need to do something like
```
assertProperties.assert(details, [], ['id', 'build_id', 'active', 'app', 'version', 'location',
                'activated_on', 'deactivated_on']);
```
but this would need more headspace to understand the place this is being called from too.

orangejenny commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -45,6 +47,7 @@ hqDefine("data_dictionary/js/data_dictionary", [
         self.caseType = caseType;
         self.dataType = ko.observable(dataType);
         self.description = ko.observable(description);
+        self.fhirRsourcePropPath = ko.observable(fhirRsourcePropPath);
```
Rsource => Resource

You're impressively consistent.

mkangia replied:
thank you. I am :)

mkangia replied:
Fixed.

orangejenny commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -17,10 +17,23 @@ <h2 class="text-hq-nav-header">{% trans "Data Dictionary" %}</h2>
+      {% if fhir_integration_enabled %}
+        <div class="form-inline" data-bind="visible: fhirResourceTypes().length">
+          {% trans "FHIR Resource Type" %}
+          <select class="form-control"
```
select2 would be nice for UI consistency, we're using them more and more places.

mkangia replied:
that would mean adding a class "hqwebapp-select2" and importing `'select2/dist/js/select2.full.min'` or something more?

orangejenny replied:
Yep, that class, but import `hqwebapp/js/widgets`, which will save you from needing to call `.select2()` yourself. There are some docs on select2s [here](https://www.commcarehq.org/styleguide/molecules/#molecules-selections), somewhat hidden in the "Select2" panel under the "Code" tab.

mkangia replied:
thanks @orangejenny 
I have made the change https://github.com/dimagi/commcare-hq/pull/29210/commits/8bb095a38bafa956b6ea7fe86ce49becc8c6c555
one thing though, now my select container is stretched how do i keep it small? i tried adding "col-sm-3" but then that lead the following items to come floating up to the right of select.

orangejenny commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -17,10 +17,23 @@ <h2 class="text-hq-nav-header">{% trans "Data Dictionary" %}</h2>
+          {% trans "FHIR Resource Type" %}
+          <select class="form-control"
+                  data-bind="options: fhirResourceTypes,
+                             optionsCaption: 'Select a resource type',
```
translate

mkangia replied:
Done, thank you

orangejenny reviewed (commented):
Just skimmed the UI code, which looks good.

mkangia committed:
update expected ID format

mkangia committed:
typo

mkangia committed:
translate

mkangia committed:
Merge branch 'master' into mk/fhir/data-dictionary

mkangia committed:
use better check

mkangia committed:
support different fhir versions

kaapstorm committed:
🔮 Future circular import

Anticipates a circular import caused when `fhir` modules use a `data_dictionary` util.

mkangia committed:
fix name

mkangia committed:
simply return 404

also ensures that the case belongs to the domain

mkangia committed:
remove fhir resource property when removed on UI

mkangia committed:
clean up

mkangia committed:
use select2

mkangia committed:
Revert "use select2"

Did not load the default value post ajax data fetch
This reverts commit 8bb095a38bafa956b6ea7fe86ce49becc8c6c555.

mkangia committed:
allow access via api key as well

mkangia committed:
extract method

mkangia committed:
better name

mkangia committed:
refactor prep to add a new sheet for case type mapping

mkangia committed:
export fhir resource mappings

kaapstorm committed:
Merge branch 'master' into mk/fhir/data-dictionary

mkangia committed:
refactor long method

mkangia committed:
update case type and prop mappings

mkangia committed:
make it explicit to remove path

mkangia committed:
add column to remove path

mkangia committed:
remove path for deprecated case properties

Deprecated properties will be ignored when sending FHIR data, so path is
redundant

mkangia committed:
remove unused import

mkangia committed:
update only when path present

mkangia committed:
stickler

mkangia committed:
remove id and type, now added by build

mkangia committed:
catch configuration errors

mkangia committed:
extract resource_url

mkangia committed:
fix circular import

mkangia committed:
use select related to get in single query

mkangia committed:
Do not fail

mkangia committed:
get resource types from first sheet instead

To get most up to date mappings

mkangia committed:
better message

mkangia committed:
update validations

mkangia committed:
validate and handle resource type

mkangia committed:
be explicit about fhir resource type errors

mkangia committed:
update case props only when no errors

mkangia committed:
Merge pull request #29336 from dimagi/mk/fhir/data-dictionary-upload-2

Bulk upload fhir mapping

mkangia committed:
Merge pull request #29335 from dimagi/mk/fhir/api-2

 Add fhir resource get/search views

mkangia committed:
move fhir specific code under fhir app

mkangia committed:
add tests for validator

mkangia committed:
add tests for utils

mkangia committed:
add ability to clear resource type mapping

mkangia committed:
isort

mkangia committed:
reset flag when switching

mkangia commented:
Hey @kaapstorm 

I think this feature is now at a good place to get merged. QA is wrapped up as well, with just the last change i have mentioned about unsetting FHIR resource type, which is not a blocker, for a pass.

Post your last review, 
1. I have just merged the other two approved PRs into this one.
2. moved some fhir specific code to fhir app as code improvement ("move fhir specific code under fhir app")
3. add tests for validation added on model and also utils. I tried reviving the view tests added in https://github.com/dimagi/commcare-hq/pull/28994/files but I ran into quite some challenges, so i am going to take that as a follow up and not block this.
4. add ability to reset fhir resource type. This was an oversight, missed it earlier that there was no way to unset fhir resource type in DD, so added that now, on same lines as we reset fhir resource case property.

Can you take a pass at the new changes or anything overall since its now at one place that you think can be improved or is out of place?

kaapstorm reviewed (approved):
looks good to me

mkangia committed:
error only when prop_path is to be updated

mkangia commented on corehq/apps/data_dictionary/templates/data_dictionary/base.html:
```diff
@@ -72,12 +100,25 @@ <h3 data-bind="text: $root.activeCaseType()"></h3>
+            <!-- ko if: deprecated() -->
+            <button data-bind="click: restoreProperty" class="btn btn-default btn-sm">{% trans "Restore Property" %}</button>
+            <!-- /ko -->
+            <!-- ko if: fhirResourcePropPath() && !removeFHIRResourcePropertyPath() -->
```
note to self: consider adding `{% if fhir_integration_enabled %}` for the buttons related to resource property path

mkangia commented on corehq/apps/data_dictionary/static/data_dictionary/js/data_dictionary.js:
```diff
@@ -59,18 +68,31 @@ hqDefine("data_dictionary/js/data_dictionary", [
         self.casePropertyList = ko.observableArray();
         self.showAll = ko.observable(false);
         self.availableDataTypes = typeChoices;
+        self.fhirResourceTypes = ko.observableArray(fhirResourceTypes);
```
note to self: these don't need to be stored as observableArray since the collection won't change.

mkangia commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -68,19 +80,26 @@ def generate_data_dictionary(request, domain):
     for case_type in queryset:
         p = {
             "name": case_type.name,
+            "fhir_resource_type": fhir_resource_type_name_by_case_type.get(case_type),
```
note to self: even though the empty key-value pair wont hurt in the json response when toggle is not enabled, consider keeping it clean and adding only when toggle enabled

mkangia commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -97,17 +116,36 @@ def data_dictionary_json(request, domain, case_type_name=None):
-        if error:
-            errors.append(error)
+
+    if request.POST.get('remove_fhir_resource_type', '') == 'true':
```
note to self: the two ifs along with the part of checking for props on request object can come in a block of if toggles.FHIR_INTEGRATION.enabled check or infact be extracted out of this method entirely

mkangia commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -128,30 +166,92 @@ def update_case_property_description(request, domain):
 
 def _export_data_dictionary(domain):
+    export_fhir_data = toggles.FHIR_INTEGRATION.enabled(domain)
+    case_type_headers = [_('Case Type'), _('FHIR Resource Type'), _('Remove Resource Type(Y)')]
```
note to self: can these be constants?

mkangia commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -238,15 +344,65 @@ def post(self, request, *args, **kwargs):
+                        if fhir_resource_prop_path and not fhir_resource_type:
+                            error = _('Could not find resource type for {}').format(case_type)
+                        else:
+                            error = save_case_property(name, case_type, domain, data_type, description, group,
```
note to self: can save_case_property be still called just once outside if block?

mkangia commented:
going to take pending improvements and self notes as a follow up.

QA is pending with just one minor improvement to be confirmed. Rest all has been tested thoroughly.
